openssl.go

```diff
@@ -34,25 +33,25 @@ var nativeEndian binary.ByteOrder
 // and if the FIPS mode is enabled.
 // This function can be called before Init.
 func CheckVersion(version string) (exists, fips bool) {
-	handle := dlopen(version)
+	handle, _ := dlopen(version)
 	if handle == nil {
 		return false, false
 	}
-	defer C.dlclose(handle)
+	defer dlclose(handle)
 	fips = C.go_openssl_fips_enabled(handle) == 1
 	return true, fips
 }
 
-// Init loads and initializes OpenSSL.
+// Init loads and initializes OpenSSL from the shared library at path.
 // It must be called before any other OpenSSL call, except CheckVersion.
 //
-// Only the first call to Init is effective,
-// subsequent calls will return the same error result as the one from the first call.
+// Only the first call to Init is effective.
+// Subsequent calls will return the same error result as the one from the first call.
 //
-// version will be appended to the OpenSSL shared library name as a version suffix
-// when calling dlopen. For example, `version=1.1.1k-fips` makes Init look for
-// the shared library libcrypto.so.1.1.1k-fips.
-func Init(version string) error {
+// The file is passed to dlopen() verbatim to load the OpenSSL shared library.
+// For example, `file=libcrypto.so.1.1.1k-fips` makes Init look for the shared
+// library libcrypto.so.1.1.1k-fips.
+func Init(file string) error {
```

Comment:
This change is not backwards compatible and will break existing users of the `openssl/v2` module.

How would everyone feel about leaving this API as-is and updating it to prepend `libcrypto.so` like `dlopen` used to, and exporting a new API like `InitFullPath` or similar? A change like this should technically wait for `v3`.

Reply:
I'm fine with changing the meaning of the `Init` parameter. Technically, `v2` hasn't been officially released yet, it is still in release candidate phase (https://github.com/golang-fips/openssl/tags), so this would be an allowed breaking change.

The Microsoft fork started to use `openssl/v2` in the current Go development cycle (go1.22), so this wouldn't affect our release branches.

Having said this, I don't have an strong opinion here. I'm also good with having something like `InitPath`.

Reply:
I suppose you're right, and we're still integrating it on our end and haven't shipped anything with it yet so not a huge deal to update our code.